### ISSUE TITLE
Support `attn_mask` in `jagged_scaled_dot_product_attention`

### ISSUE THREAD

**Issue by schmidt-jake** [pytorch/pytorch#138993]
### 🚀 The feature, motivation and pitch

Currently, providing an attention mask argument to [jagged_scaled_dot_product_attention](https://github.com/pytorch/pytorch/blob/main/torch/nested/_internal/sdpa.py#L677) is not supported (see https://github.com/pytorch/pytorch/blob/main/torch/nested/_internal/sdpa.py#L67).

I don't know how technically feasibly this is, as I'm not familiar with the structure of the various SDPA backends, but theoretically I'm hoping we could allow passing a nested tensor attn_mask (strided layout only, not jagged, b/c by definition the nested tensor will have two ragged dimensions).

### Alternatives

For use cases where attention masks are required but inputs are jagged, I believe this only leaves the option of converting everything to a dense padded tensor, which is not very efficient.

**Comment by vulkomilev** [comment]
I will look at it

**Comment by cpuhrsch** [comment]
You may also like https://github.com/pytorch/pytorch/pull/136792 for https://pytorch.org/blog/flexattention/ support for NJT

**Comment by jbschlosser** [comment]
As @cpuhrsch mentioned, FlexAttention is a more flexible way to get this `attn_mask` support without the need for strided layout NT attn_masks, and luckily the PR just landed :) I'd suggest trying this out if you're willing to and reporting any issues you run into here.

**Comment by schmidt-jake** [comment]
Thanks @cpuhrsch and @jbschlosser! I'll try that out.

**Comment by schmidt-jake** [comment]
Ok @cpuhrsch and @jbschlosser, I've tried out `create_nested_block_mask` from #136792! It looks really promising. I ran into difficulty only because I think my use case falls outside the provided examples I've found for flex attention ([here](https://pytorch.org/blog/flexattention/) and [here](https://github.com/pytorch-labs/attention-gym/tree/main)). I'm hoping you might enlighten me...

My use case is effectively "nested" document masking with global tokens. So what this means is a batch is nested but not jagged — it has the shape (batch_size, *num_docs, *doc_length, ...) where a `*` means ragged. The tokens within a document can attend to the global tokens and each other, but not to tokens in other documents. The global tokens can attend to all tokens. Here's a drawing of what the attention pattern should look like for batch indices 0 and 1.

![image](https://github.com/user-attachments/assets/88c88c4c-5521-4199-b2a3-bca7ad5684f0)

I think I've solved the global attention part with this:

```python
from torch.nn.attention.flex_attention import create_nested_block_mask, or_masks


def global_q_mask_mod(
    b: torch.Tensor, h: torch.Tensor, q_idx: torch.Tensor, kv_idx: torch.Tensor, n: int
) -> torch.Tensor:
    return q_idx < n


def global_kv_mask_mod(
    b: torch.Tensor, h: torch.Tensor, q_idx: torch.Tensor, kv_idx: torch.Tensor, n: int
) -> torch.Tensor:
    return kv_idx < n

mask_mod = or_masks(global_q_mask_mod, global_kv_mask_mod)
mask = create_nested_block_mask(mask_mod, ...)
```

In practice, I reshape the batch to (batch_size, *num_tokens, ...), making it jagged, and maintain a separate nested tensor to track the document IDs of each token.

The problem is how to create the document masking logic. I was hoping to do something like this:

```python
def document_mask(document_ids: NestedTensor) -> _mask_mod_signature:
    def mask_mod(b, h, q_idx, kv_idx):
        _doc_ids = document_ids[b] # the doc IDs for this example in the batch
       return _docs_ids[q_idx] == _docs_ids[kv_id]
    return mask_mod
```

But this will throw a `NotImplementedError: aten.index.Tensor`. So now my questions are:
1. Am I abusing the feature?
2. If not, is there a better pattern to use here, or is there a piece of functionality I need to implement to make this work?

Sorry for the long post. Any advice is very appreciated!

**Comment by jbschlosser** [comment]
Hey @schmidt-jake, thanks for the detailed report! Sorry for the delay - I was out on PTO.

> Am I abusing the feature?

Certainly not, this seems like a valid use case and exactly the sort of thing we should support :) Thanks for the detailed diagrams here illuminating what you're trying to do!

> If not, is there a better pattern to use here, or is there a piece of functionality I need to implement to make this work?

On the NJT side, we should support the ops that are blocking you; in this case, tensor indexing on the batch dimension stemming from `document_ids[b]` and `_doc_ids[q_idx]` / `_doc_ids[kv_idx]`. I started an implementation for this, and it's tricky for two reasons:
1. The output of tensor indexing on the batch dim of an NJT has a data-dependent shape, which are difficult to handle in torch.compile (some related [context](https://github.com/pytorch/pytorch/pull/136444#issuecomment-2430128489)).
2. FlexAttention uses vmap underneath, and your (valid) approach of utilizing an NJT as a buffer to store doc IDs results in vmap + NJT interaction, which sadly doesn't work today, but we do really want this.

That said, let me chat with other NJT folks and see if we can come up with some workaround for you to accomplish what you need  before these issues are fixed.

**Comment by schmidt-jake** [comment]
Fantastic! I wound up hacking around the indexing issue:

```python
def nt_index(x: torch.Tensor, b: torch.Tensor, ix: torch.Tensor) -> torch.Tensor:
    offset = x._offsets[b]
    return x._values[offset + ix, ...]
```

A related-but-maybe-separate issue is NJT cross attention support. Should I open a separate issue?

**Comment by jbschlosser** [comment]
> A related-but-maybe-separate issue is NJT cross attention support. Should I open a separate issue?

Sure, that'd be great! Thanks @schmidt-jake :)

**Comment by jbschlosser** [comment]
@schmidt-jake One workaround that should unblock you today is to materialize a padded tensor for the document ID level of raggedness, while keeping NJTs for the doc length level of raggedness.
```python
...
# convert document_ids from NJT -> padded dense.
# index.Tensor / vmap support for NJT is needed to avoid this
document_ids = document_ids.to_padded_tensor(-1)

def my_mask_mod(b, h, q_idx, kv_idx):
    doc_ids = document_ids[b]
    # filter out padding doc ids
    return (doc_ids[q_idx] == doc_ids[kv_idx]) & (doc_ids[q_idx] > -1)

block_mask = create_nested_block_mask(my_mask_mod, 1, 1, query, _compile=True)
flex_attention(query, key, value, block_mask=block_mask)
```

Obviously, this is non-ideal performance-wise since this full padded tensor is materialized. After we add the support I mentioned above, you should be able to avoid this materialization and get better perf.
